File: backend/voidswitch/api/proxy.py
"""Public gateway endpoints — the OpenAI- and Anthropic-style inbound APIs.

Point an OpenAI client at ``/v1/chat/completions`` (Chat Completions) or
``/v1/responses`` (the Responses API), or Claude Code / an Anthropic client at
``/v1/messages`` (set ANTHROPIC_BASE_URL to this server). Inbound and upstream
styles are translated transparently by the dispatcher.
"""

from __future__ import annotations

import datetime as dt
import re
from fnmatch import fnmatch

from fastapi import APIRouter, Depends, Header, HTTPException, Request, status
from fastapi.responses import JSONResponse, Response, StreamingResponse
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from voidswitch.constants import (
    CLIENT_HINT_HEADER,
    OPENCODE_CLIENT_HINT,
    SESSION_HEADER,
    UPSTREAM_UNAVAILABLE_STATUS,
    ApiStyle,
)
from voidswitch.core import auth, ratelimit
from voidswitch.core.database import get_session
from voidswitch.core.logging import get_logger, redact_headers
from voidswitch.models.db import ExposedModel, Provider, RequestLog, VoidToken
from voidswitch.services import model_routing, role_groups, settings_store
from voidswitch.services.dispatcher import DispatchRequest, dispatch

router = APIRouter(tags=["gateway"])
log = get_logger("gateway")


_PASSTHROUGH_RE = re.compile(
    r"^(?P<exposed>[^\s@]+?)(?:\s*=>\s*(?P<upstream>[^\s@]+?))?(?:\s*@\s*(?P<pool>\S+))?$"
)


def _parse_passthrough_entry(entry: str) -> dict[str, str]:
    """Parse a passthrough whitelist entry into its components."""
    m = _PASSTHROUGH_RE.match(entry.strip())
    if m is None:
        return {"exposed": entry.strip(), "upstream": entry.strip(), "pool": ""}
    exposed = (m.group("exposed") or "").strip()
    upstream = (m.group("upstream") or exposed).strip()
    pool = (m.group("pool") or "").strip()
    return {"exposed": exposed, "upstream": upstream, "pool": pool}


def _check_rpm(token: VoidToken) -> None:
    """Enforce a Void-Token's per-minute request cap (sliding 60s window).

    Backed by the shared, single-node :data:`ratelimit.gateway_rpm_limiter` — see
    that module's note on the multi-worker caveat (counters are per-process).
    """
    if token.rpm_limit <= 0:
        return
    if not ratelimit.gateway_rpm_limiter.allow(
        f"rpm:{token.id}", window_seconds=60.0, max_requests=token.rpm_limit
    ):
        raise HTTPException(
            status.HTTP_429_TOO_MANY_REQUESTS,
            f"Rate limit exceeded ({token.rpm_limit} req/min).",
        )


async def _check_daily_quota(session: AsyncSession, token: VoidToken) -> None:
    """Enforce a Void-Token's per-UTC-day request cap (``daily_quota``).

    Counts today's logged requests for the token (the same rows the request-log
    browser shows). 0 = unlimited.
    """
    if token.daily_quota <= 0:
        return
    start_of_day = dt.datetime.now(dt.UTC).replace(hour=0, minute=0, second=0, microsecond=0)
    used = (
        await session.execute(
            select(func.count(RequestLog.id)).where(
                RequestLog.token_id == token.id, RequestLog.ts >= start_of_day
            )
        )
    ).scalar_one()
    if used >= token.daily_quota:
        raise HTTPException(
            status.HTTP_429_TOO_MANY_REQUESTS,
            f"Daily request quota exceeded ({token.daily_quota} requests/day).",
        )


def _check_call_rate_limit(user_id: int) -> None:
    """Per-user abuse limit on the OpenAI/Anthropic gateway endpoints.

    Independent of a token's own ``rpm_limit`` — this is a platform-wide guard
    that everyone obeys (owners included), counted per user. Disabled when the
    configured max is 0.
    """
    window = settings_store.get_int("call_rate_limit_window_seconds", 60)
    max_requests = settings_store.get_int("call_rate_limit_max_requests", 0)
    if not ratelimit.call_limiter.allow(
        f"call:{user_id}", window_seconds=window, max_requests=max_requests
    ):
        raise HTTPException(
            status.HTTP_429_TOO_MANY_REQUESTS,
            f"Call rate limit exceeded ({max_requests} per {window}s). Slow down.",
        )


def _check_model_allowed(token: VoidToken, model: str) -> None:
    allowed = token.allowed_models or []
    if not allowed:
        return
    for pattern in allowed:
        if pattern == "*" or pattern == model or fnmatch(model, pattern):
            return
    raise HTTPException(
        status.HTTP_403_FORBIDDEN, f"Model '{model}' is not permitted for this token."
    )


async def _resolve_public_model(session: AsyncSession, model: str) -> str:
    """Map a public model id to its exposed-model row.

    Only *exposed* ids are callable. Raw upstream ids (``slug/model``) are never
    advertised and are rejected here, so an upstream id can't be reached
    directly.

    Supports passthrough: when the model is in ``provider-slug/exposed-model-id``
    format, the provider is looked up and its passthrough whitelist is checked.
    """
    parts = model.split("/", 1)
    if len(parts) == 2:
        provider_slug = parts[0]
        exposed_id = parts[1]
        provider = (
            await session.execute(select(Provider).where(Provider.slug == provider_slug))
        ).scalar_one_or_none()
        if provider is not None and provider.enabled and provider.passthrough_enabled:
            for entry in provider.passthrough_models or []:
                parsed = _parse_passthrough_entry(entry)
                if parsed["exposed"] == exposed_id:
                    return model
            raise HTTPException(
                status.HTTP_404_NOT_FOUND,
                f"Model '{model}' is not in this provider's passthrough list.",
            )
    entry = (
        await session.execute(select(ExposedModel).where(ExposedModel.model_id == model))
    ).scalar_one_or_none()
    if entry is None:
        raise HTTPException(
            status.HTTP_404_NOT_FOUND,
            f"Model '{model}' is not available under this id.",
        )
    if not entry.enabled:
        raise HTTPException(
            status.HTTP_404_NOT_FOUND,
            f"Model '{model}' is not available under this id.",
        )
    return entry.model_id


async def _body(request: Request) -> dict:
    try:
        data = await request.json()
    except Exception as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Invalid JSON body.") from exc
    if not isinstance(data, dict):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Request body must be a JSON object.")
    return data


def _client_ip(request: Request) -> str | None:
    fwd = request.headers.get("x-forwarded-for")
    if fwd:
        return fwd.split(",")[0].strip()
    return request.client.host if request.client else None


async def _handle(
    request: Request,
    session: AsyncSession,
    authorization: str | None,
    x_api_key: str | None,
    inbound_style: ApiStyle,
) -> Response:
    authed = await auth.authenticate_void_token(session, authorization, x_api_key)
    # Platform-wide per-user call rate limit (abuse guard, everyone incl. owners).
    _check_call_rate_limit(authed.user.id)
    payload = await _body(request)

    model = payload.get("model")
    if not isinstance(model, str) or not model:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Missing 'model'.")

    # A request body in the wrong shape (e.g. ``messages`` a string instead of a
    # list) would crash the translator later as a 500 after upstream billing;
    # reject it up front as a 400.
    for field in ("messages", "tools", "input"):
        if (
            field in payload
            and payload[field] is not None
            and not isinstance(payload[field], (list, str) if field == "input" else list)
        ):
            raise HTTPException(status.HTTP_400_BAD_REQUEST, f"'{field}' must be a list.")

    # Allow-list and rate-limit are checked against the *public* id the caller
    # used, then the model is resolved to its real upstream id (if it's an alias).
    _check_model_allowed(authed.token, model)
    _check_rpm(authed.token)
    await _check_daily_quota(session, authed.token)
    model = await _resolve_public_model(session, model)
    payload["model"] = model

    # Role-group access: a non-moderator may only call models whose allowed role
    # groups intersect their own. Moderators may call everything. Passthrough
    # models (``provider-slug/exposed-model-id``) are whitelist-controlled by the
    # provider and skip the role-group check.
    if "/" not in model and not await role_groups.user_can_access_model(
        session, authed.user, model
    ):
        raise HTTPException(
            status.HTTP_403_FORBIDDEN,
            f"Model '{model}' is not available to your role group.",
        )

    stream = bool(payload.get("stream", False))
    passthrough: dict[str, str] = {}
    beta = request.headers.get("anthropic-beta")
    if beta and inbound_style is ApiStyle.ANTHROPIC:
        passthrough["anthropic-beta"] = beta

    # Verbose inbound tracing (debug mode only — gated by the log level so it
    # costs nothing when disabled). Credentials in the headers are masked.
    log.debug(
        "inbound_request",
        method=request.method,
        path=request.url.path,
        inbound_style=inbound_style.value,
        model=model,
        stream=stream,
        token_id=authed.token.id,
        client_ip=_client_ip(request),
        headers=redact_headers(request.headers),
        body=payload,
    )

    req = DispatchRequest(
        inbound_style=inbound_style,
        model=model,
        payload=payload,
        stream=stream,
        token_id=authed.token.id,
        user_sub=authed.user.sub,
        client_ip=_client_ip(request),
        user_agent=request.headers.get("user-agent"),
        client_type=request.headers.get(CLIENT_HINT_HEADER),
        is_opencode=request.headers.get(CLIENT_HINT_HEADER) == OPENCODE_CLIENT_HINT,
        debug_enabled=authed.token.debug_enabled,
        passthrough_headers=passthrough,
        session_id=request.headers.get(SESSION_HEADER),
    )
    result = await dispatch(req, session=session)

    if result.is_stream and result.stream is not None:
        return StreamingResponse(
            result.stream,
            status_code=result.status_code,
            media_type=result.media_type,
            headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
        )

    status_code = result.status_code
    # When no upstream could serve the request, swap the generic 502 ("Bad Gateway")
    # for a dedicated code — but only for the OpenCode plugin, which advertises itself
    # and knows how to render it as "Upstream Failed". Every other client keeps the
    # standard 502 so SDKs and intermediaries aren't surprised by a non-standard code.
    if (
        result.error == "upstream_unavailable"
        and request.headers.get(CLIENT_HINT_HEADER) == OPENCODE_CLIENT_HINT
    ):
        status_code = UPSTREAM_UNAVAILABLE_STATUS

    return Response(
        content=result.content or b"{}",
        status_code=status_code,
        media_type=result.media_type,
    )


@router.post("/v1/chat/completions")
async def chat_completions(
    request: Request,
    session: AsyncSession = Depends(get_session),
    authorization: str | None = Header(default=None),
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
) -> Response:
    return await _handle(request, session, authorization, x_api_key, ApiStyle.OPENAI)


@router.post("/v1/messages")
async def messages(
    request: Request,
    session: AsyncSession = Depends(get_session),
    authorization: str | None = Header(default=None),
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
) -> Response:
    return await _handle(request, session, authorization, x_api_key, ApiStyle.ANTHROPIC)


@router.post("/v1/responses")
async def responses(
    request: Request,
    session: AsyncSession = Depends(get_session),
    authorization: str | None = Header(default=None),
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
) -> Response:
    return await _handle(request, session, authorization, x_api_key, ApiStyle.OPENAI_RESPONSES)


async def _advertised_models(
    session: AsyncSession, authed: auth.AuthedToken
) -> list[dict[str, object]]:
    """Models this caller may actually use, in ``/v1/models`` payload shape.

    Only *exposed* models are advertised (upstream ``slug/model`` ids never are).
    Role-group access is enforced (moderators see everything; others only models
    their groups allow), and the token's ``allowed_models`` allow-list is
    honoured. The ``opencode`` block is the fully merged downstream config
    (defaults ← models.dev placeholder ← custom config ← structured fields).

    Passthrough models from providers with ``passthrough_enabled=True`` are also
    listed as ``provider-slug/exposed-model-id``.
    """
    token = authed.token

    entries = (await session.execute(select(ExposedModel))).scalars().all()

    is_mod = role_groups.is_moderator(authed.user)
    group_ids = set() if is_mod else await role_groups.user_group_ids(session, authed.user.id)
    seen: set[str] = set()
    data: list[dict[str, object]] = []
    allowed = token.allowed_models or []

    # Pre-load the models.dev entries referenced by any exposed model, once.
    dev_ids = [e.models_dev_id for e in entries if e.models_dev_id]
    md_by_id: dict[str, dict] = {}
    if dev_ids:
        from sqlalchemy import select as _select

        from voidswitch.models.db import ModelsDevCache

        rows = (
            (await session.execute(_select(ModelsDevCache).where(ModelsDevCache.id.in_(dev_ids))))
            .scalars()
            .all()
        )
        md_by_id = {r.id: r.data or {} for r in rows}

    def _push(entry: ExposedModel) -> None:
        nonlocal data, seen, allowed, group_ids, is_mod, md_by_id
        if not entry.enabled:
            return
        if not role_groups.model_allowed_for_groups(entry, group_ids, is_mod=is_mod):
            return
        model_id = entry.model_id
        if model_id in seen:
            return
        if allowed and not any(
            pattern == "*" or pattern == model_id or fnmatch(model_id, pattern)
            for pattern in (str(p) for p in allowed)
        ):
            return
        seen.add(model_id)
        item: dict[str, object] = {
            "id": model_id,
            "object": "model",
            "created": 0,
            "owned_by": "voidswitch",
        }
        # Downstream config: never advertise upstream refs.
        dev_entry = md_by_id.get(entry.models_dev_id) if entry.models_dev_id else None
        item["opencode"] = model_routing.build_opencode_config(entry, dev_entry)
        if entry.display_name:
            item["display_name"] = entry.display_name
        if entry.description:
            item["description"] = entry.description
        data.append(item)

    for entry in entries:
        _push(entry)

    # Passthrough models: each provider with passthrough_enabled contributes its
    # whitelisted models as ``provider-slug/exposed-model-id``.
    passthrough_providers = (
        (
            await session.execute(
                select(Provider).where(
                    Provider.enabled.is_(True), Provider.passthrough_enabled.is_(True)
                )
            )
        )
        .scalars()
        .all()
    )
    for provider in passthrough_providers:
        for entry in provider.passthrough_models or []:
            parsed = _parse_passthrough_entry(entry)
            exposed_id = parsed["exposed"]
            model_id = f"{provider.slug}/{exposed_id}"
            if model_id in seen:
                continue
            if allowed and not any(
                pattern == "*" or pattern == model_id or fnmatch(model_id, pattern)
                for pattern in (str(p) for p in allowed)
            ):
                continue
            seen.add(model_id)
            data.append(
                {
                    "id": model_id,
                    "object": "model",
                    "created": 0,
                    "owned_by": "voidswitch",
                }
            )

    return data


@router.get("/v1/models")
async def list_models(
    session: AsyncSession = Depends(get_session),
    authorization: str | None = Header(default=None),
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
) -> JSONResponse:
    authed = await auth.authenticate_void_token(session, authorization, x_api_key)
    data = await _advertised_models(session, authed)
    return JSONResponse({"object": "list", "data": data})


@router.post("/v1/models/sync")
async def sync_models(
    session: AsyncSession = Depends(get_session),
    authorization: str | None = Header(default=None),
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
) -> JSONResponse:
    """Report the models the caller can use (OpenCode ``/sync-models`` command).

    Open to **any** authenticated Void-Token, members included: it returns the
    exact set of models that token may call — role-group access and the token's
    allow-list applied, hidden/disabled models excluded — so the OpenCode plugin
    can refresh its provider model map to what the *user* actually has.

    It deliberately does **not** reshape the shared platform catalog; that is the
    staff-only "sync from providers" action on the dashboard
    (``POST /api/models/sync``). Keeping the two apart means a member's sync never
    needs admin rights. Also returns the gateway's recommended OpenCode
    ``model`` / ``small_model`` selectors so the plugin can sync the top-level
    config keys alongside the provider's model map.
    """
    authed = await auth.authenticate_void_token(session, authorization, x_api_key)
    data = await _advertised_models(session, authed)
    return JSONResponse(
        {
            "added": 0,
            "total": len(data),
            "opencode_default_model": settings_store.get_str(
                "opencode_default_model", "claude-opus-4-8"
            ),
            "opencode_small_model": settings_store.get_str("opencode_small_model", ""),
        }
    )

File: backend-go/internal/handlers/proxy.go
package handlers

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/siiway/voidswitch/internal/constants"
	"github.com/siiway/voidswitch/internal/core"
	"github.com/siiway/voidswitch/internal/database"
	"github.com/siiway/voidswitch/internal/services"
)

const Version = "0.1.0"

func HandleRoot(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"name":    "VoidSwitch",
		"version": Version,
		"endpoints": []string{
			"/",
			"/healthz",
			"/v1/chat/completions",
			"/v1/messages",
			"/v1/models",
			"/v1/models/sync",
		},
	})
}

func HandleHealthz(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"status":  "ok",
		"version": Version,
	})
}

func HandleChatCompletions(c *gin.Context) {
	db := database.GetDatabase()

	authHeader := c.GetHeader("Authorization")
	xApiKey := c.GetHeader("X-API-Key")
	raw := extractBearer(authHeader)
	if raw == "" {
		raw = strings.TrimSpace(xApiKey)
	}
	if raw == "" {
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": gin.H{
				"message": "Missing API key.",
				"type":    "auth_error",
				"code":    "auth_error",
			},
		})
		return
	}

	authed, err := core.AuthenticateVoidToken(db.DB, authHeader, xApiKey)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": gin.H{
				"message": fmt.Sprintf("Invalid API key: %v", err),
				"type":    "auth_error",
				"code":    "auth_error",
			},
		})
		return
	}

	var payload map[string]any
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": gin.H{
				"message": "Invalid request body.",
				"type":    "invalid_request_error",
			},
		})
		return
	}

	stream := false
	if s, ok := payload["stream"]; ok {
		if b, ok := s.(bool); ok {
			stream = b
		}
	}

	model, _ := payload["model"].(string)
	userSub := authed.User.Sub

	req := &services.DispatchRequest{
		InboundStyle: constants.ApiStyleOpenAI,
		Model:        model,
		Payload:      payload,
		Stream:       stream,
		TokenID:      &authed.Token.ID,
		UserSub:      &userSub,
	}

	log.Printf("[proxy] chat/completions model=%s stream=%v token=%d user=%s",
		model, stream, authed.Token.ID, core.ActorDisplayName(authed.User))

	result, _ := services.Dispatch(req)

	if result.IsStream {
		writeSSE(c, result.StreamCh)
		return
	}

	c.Data(result.StatusCode, result.MediaType, result.Content)
}

func HandleMessages(c *gin.Context) {
	db := database.GetDatabase()

	authHeader := c.GetHeader("Authorization")
	xApiKey := c.GetHeader("X-API-Key")

	authed, err := core.AuthenticateVoidToken(db.DB, authHeader, xApiKey)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": gin.H{
				"type":    "authentication_error",
				"message": fmt.Sprintf("Invalid API key: %v", err),
			},
		})
		return
	}

	var payload map[string]any
	if err := c.ShouldBindJSON(&payload); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": gin.H{
				"type":    "invalid_request_error",
				"message": "Invalid request body.",
			},
		})
		return
	}

	stream := false
	if s, ok := payload["stream"]; ok {
		if b, ok := s.(bool); ok {
			stream = b
		}
	}

	model, _ := payload["model"].(string)
	userSub := authed.User.Sub

	req := &services.DispatchRequest{
		InboundStyle: constants.ApiStyleAnthropic,
		Model:        model,
		Payload:      payload,
		Stream:       stream,
		TokenID:      &authed.Token.ID,
		UserSub:      &userSub,
	}

	log.Printf("[proxy] messages model=%s stream=%v token=%d user=%s",
		model, stream, authed.Token.ID, core.ActorDisplayName(authed.User))

	result, _ := services.Dispatch(req)

	if result.IsStream {
		writeSSE(c, result.StreamCh)
		return
	}

	c.Data(result.StatusCode, result.MediaType, result.Content)
}

func HandleModels(c *gin.Context) {
	db := database.GetDatabase()

	authHeader := c.GetHeader("Authorization")
	xApiKey := c.GetHeader("X-API-Key")

	_, err := core.AuthenticateVoidToken(db.DB, authHeader, xApiKey)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": gin.H{
				"message": fmt.Sprintf("Invalid API key: %v", err),
				"type":    "auth_error",
				"code":    "auth_error",
			},
		})
		return
	}

	catalog, err := services.BuildCatalog(db.DB)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": gin.H{
				"message": fmt.Sprintf("Failed to build catalog: %v", err),
				"type":    "internal_error",
			},
		})
		return
	}

	now := time.Now().Unix()

	models := make([]gin.H, 0, len(catalog))
	for _, m := range catalog {
		created := now
		if m.CreatedAt != nil {
			created = m.CreatedAt.Unix()
		}

		models = append(models, gin.H{
			"id":       m.ModelID,
			"object":   "model",
			"created":  created,
			"owned_by": "voidswitch",
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"object": "list",
		"data":   models,
	})
}

func HandleModelsSync(c *gin.Context) {
	db := database.GetDatabase()

	authHeader := c.GetHeader("Authorization")
	xApiKey := c.GetHeader("X-API-Key")

	authed, err := core.AuthenticateVoidToken(db.DB, authHeader, xApiKey)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"error": gin.H{
				"message": fmt.Sprintf("Invalid API key: %v", err),
				"type":    "auth_error",
				"code":    "auth_error",
			},
		})
		return
	}

	actorName := core.ActorDisplayName(authed.User)

	result, err := services.SyncCatalog(db.DB, &authed.User.ID, &actorName)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": gin.H{
				"message": fmt.Sprintf("Failed to sync catalog: %v", err),
				"type":    "internal_error",
			},
		})
		return
	}

	log.Printf("[proxy] models/sync added=%d total=%d actor=%s", result.Added, result.Total, actorName)

	c.JSON(http.StatusOK, result)
}

func extractBearer(authorization string) string {
	if authorization == "" {
		return ""
	}
	parts := strings.SplitN(authorization, " ", 2)
	if len(parts) == 2 && strings.EqualFold(parts[0], "bearer") {
		return strings.TrimSpace(parts[1])
	}
	return strings.TrimSpace(authorization)
}

func writeSSE(c *gin.Context, ch <-chan []byte) {
	c.Writer.Header().Set("Content-Type", "text/event-stream")
	c.Writer.Header().Set("Cache-Control", "no-cache")
	c.Writer.Header().Set("Connection", "keep-alive")
	c.Writer.WriteHeader(http.StatusOK)

	flusher, ok := c.Writer.(http.Flusher)
	if !ok {
		return
	}

	for chunk := range ch {
		_, err := c.Writer.Write(chunk)
		if err != nil {
			return
		}
		flusher.Flush()
	}

	_, _ = io.WriteString(c.Writer, "data: [DONE]\n\n")
	flusher.Flush()
}

func RegisterProxyRoutes(router *gin.RouterGroup) {
	router.GET("/", HandleRoot)
	router.GET("/healthz", HandleHealthz)
	router.POST("/v1/chat/completions", HandleChatCompletions)
	router.POST("/v1/messages", HandleMessages)
	router.GET("/v1/models", HandleModels)
	router.POST("/v1/models/sync", HandleModelsSync)
}
